File: cpp/questions/sumFourDivisors.cpp
class Solution {
private:
    int fourDivisorsSum(const int& num) {
        if (num < 6) {
            return 0;
        }
        int sqrt_num = static_cast<int>(sqrt(num));
        if (num == sqrt_num * sqrt_num) {
            return 0;
        }
        int sum = 1+num;
        for (int n=2; n<=sqrt_num; n++) {
            if (num % n == 0) {
                if (sum == num+1) {
                    sum += n+num/n;
                } else {
                    return 0;
                }
            }
        }
        return sum==1+num ? 0 : sum;
    }
public:
    int sumFourDivisors(vector<int>& nums) {
        int ans = 0;
        std::map<int, int> sm;
        for (const int& num : nums) {
            auto it = sm.find(num);
            if (it != sm.end()) {
                ans += sm[num];
            } else {
                sm[num] = fourDivisorsSum(num);
                ans += sm[num];
            }
        }
        return ans;
    }
};

File: python/questions/sumFourDivisors.py
class Solution:
    def sumFourDivisors(self, nums: List[int]) -> int:
        def fourDivisorsSum(num: int) -> int:
            if num < 6:
                return 0
            sqrt_num = int(num**0.5)
            if sqrt_num * sqrt_num == num:
                return 0
            sum = 1+num
            for n in range(2, sqrt_num+1):
                if num % n == 0:
                    if sum != 1+num:
                        return 0
                    sum += n+num//n
            if sum == 1+num:
                return 0
            return sum
        csd = {}
        ans = 0
        for num in nums:
            if csd.get(num, -1) == -1:
                csd[num] = fourDivisorsSum(num)
                ans += csd[num]
            else:
                ans += csd[num]
        return ans
